tfprotov5/tf5server/server.go

```diff
@@ -130,61 +184,170 @@ func (s *server) stoppableContext(ctx context.Context) context.Context {
+	ctx = tflog.With(ctx, "tf_provider_addr", s.name)
+	return ctx
+}
+
+func rpcLoggingContext(ctx context.Context, rpc string) context.Context {
+	ctx = tfsdklog.With(ctx, "tf_rpc", rpc)
+	ctx = tfsdklog.SubsystemWith(ctx, tflogSubsystemName, "tf_rpc", rpc)
+	ctx = tflog.With(ctx, "tf_rpc", rpc)
+	return ctx
+}
+
+func resourceLoggingContext(ctx context.Context, resource string) context.Context {
+	ctx = tfsdklog.With(ctx, "tf_resource_type", resource)
+	ctx = tfsdklog.SubsystemWith(ctx, tflogSubsystemName, "tf_resource_type", resource)
+	ctx = tflog.With(ctx, "tf_resource_type", resource)
+	return ctx
+}
+
+func dataSourceLoggingContext(ctx context.Context, dataSource string) context.Context {
+	ctx = tfsdklog.With(ctx, "tf_data_source_type", dataSource)
+	ctx = tfsdklog.SubsystemWith(ctx, tflogSubsystemName, "tf_data_source_type", dataSource)
+	ctx = tflog.With(ctx, "tf_data_source_type", dataSource)
+	return ctx
+}
+
 // New converts a tfprotov5.ProviderServer into a server capable of handling
 // Terraform protocol requests and issuing responses using the gRPC types.
-func New(serve tfprotov5.ProviderServer) tfplugin5.ProviderServer {
+func New(name string, serve tfprotov5.ProviderServer, opts ...ServeOpt) tfplugin5.ProviderServer {
+	var conf ServeConfig
+	for _, opt := range opts {
+		err := opt.ApplyServeOpt(&conf)
+		if err != nil {
+			// this should never happen, we already executed all
+			// this code as part of Serve
+			panic(err)
+		}
+	}
+	var sdkOptions tfsdklog.Options
+	var options tflog.Options
+	sdkOptions = append(sdkOptions, tfsdklog.WithLevelFromEnv("TF", "LOG", "SDK", "PROTO"))
+	if !conf.disableLogInitStderr {
+		sdkOptions = append(sdkOptions, tfsdklog.WithStderrFromInit())
+		options = append(options, tfsdklog.WithStderrFromInit())
+	}
+	if conf.disableLogLocation {
+		sdkOptions = append(sdkOptions, tfsdklog.WithoutLocation())
+		options = append(options, tflog.WithoutLocation())
+	}
+	envVar := conf.envVar
+	if envVar == "" {
+		addr, err := tfaddr.ParseRawProviderSourceString(name)
+		if err != nil {
+			log.Println("[ERROR] Error parsing provider name:", err)
+		} else {
+			envVar = addr.Type
+		}
+	}
+	envVar = strings.ReplaceAll(envVar, "-", "_")
+	if envVar != "" {
+		options = append(options, tfsdklog.WithLogName(envVar), tflog.WithLevelFromEnv("TF", "LOG", "PROVIDER", envVar))
+	}
 	return &server{
-		downstream: serve,
-		stopCh:     make(chan struct{}),
+		downstream:   serve,
+		stopCh:       make(chan struct{}),
+		tflogOpts:    options,
+		tflogSDKOpts: sdkOptions,
+		name:         name,
 	}
 }
 
 func (s *server) GetSchema(ctx context.Context, req *tfplugin5.GetProviderSchema_Request) (*tfplugin5.GetProviderSchema_Response, error) {
+	ctx = rpcLoggingContext(s.loggingContext(ctx), "GetSchema")
 	ctx = s.stoppableContext(ctx)
+	tfsdklog.SubsystemTrace(ctx, tflogSubsystemName, "Received request")
+	defer tfsdklog.SubsystemTrace(ctx, tflogSubsystemName, "Served request")
```

Comment:
Open question: Should these trace level logs include the actual request/response?

Reply:
So my hot take is "no" _here_, specifically. The rationale being that request/response gRPC bodies are almost certainly going to include sensitive information, and we've gotten practitioners into the habit of supplying TRACE level logs when opening issues. So if they don't realize what they're sharing, that way lies sadness. Also, we don't want to make people hesitant to give us logs because they have so much data in them.

However, I'd be in favor of a separate sublogger like `TF_LOG_SDK_REQ_DATA` or something that could log request/response bodies at the TRACE level, but defaults to off unless it _explicitly_ is turned on. So we could opt into having request/response bodies logged, but they don't turn on by default just because you turn on all the other logging.

Reply:
Sounds very reasonable to me. Created https://github.com/hashicorp/terraform-plugin-go/issues/107 for future consideration. 👍